### PR TITLE
Allow for Input source security in SQL layer

### LINKED ISSUE #13837: Input source security model for MSQ table functions and more
MSQ provides a powerful way to ingest data into Druid using SQL, placing ingestion in the hands of more users. Prior to MSQ, only those users with sufficient knowledge to use a batch ingestion spec could ingest data. With the wider audience comes the need to add protections to prevent users from accessing data that doesn't make sense in a particular Druid deployment. This proposal outlines a proposed extension to the Druid security model to address this issue. As it turns out the Druid catalog feature has similar needs, which we also address.

## Input Source Security in Druid Today

Let us start by reviewing the current security model. Druid's model has two parts:

* A "resource action" (class `ResourceAction`) which is essentially a triple of (category, name, read/write).
* An "authenticator mapper (class `AuthenticatorMapper`) which provides a yes/no answer to the question, "does the user have access to this set of resource actions"?

In MSQ today, there is a single resource action: `(EXTERNAL, EXTERNAL, READ)` which was introduced along with the `extern` table function. That is, that single permission allows the user to read data from S3, from the local disk, from HTTP, etc.

The key gap in this model is the lack of fine-grain control over which input sources to allow. (A site may want to allow access to S3, but not the local file system on each ingest node.)

## Proposed Input Source Security Model

The proposed model extends the current system to add additional items to the second (name) element of the resource action. Specifically, to use the JSON type name for each input source, so that access to the HTTP input source would be checked as `(EXTERNAL, http, READ)`. Similarly for all other Druid input sources.

MSQ currently provides multiple ways to access a given input source: via the `extern` function, or via the newer input-source-specific functions such as `http` or `localfiles`. By applying security at the level of the input source, the same security rules apply regardless of the function used to access the input source.

## Security Model for External Tables Defined in the Catalog

The Druid catalog provides the ability to define an external table: a metadata entry that can be used in MSQ queries in lieu of spelling out the input source details in each query. The catalog already uses security rules that mimic Druid datasources. To access an external table the user needs permission on `(ext, <table name>, READ)`, where `ext` is a new Druid schema introduced to hold external table definitions.

We propose to extend the security model to _also_ need permission on the underlying input source type. Thus, if `myS3` is an external table that accesses S3, then the user needs both `(ext, myS3, READ)` and `(EXTERN, s3, READ)` permissions. 

## Security Model for Custom MSQ Table Functions

A great strength of Druid is the ability to add new input sources and "user defined" SQL functions. These can be combined to provide new MSQ table functions for new input sources. The simplest such function is a wrapper. Suppose I want to ingest data from [IPFS](https://en.wikipedia.org/wiki/InterPlanetary_File_System). I would first write the input source, then provide an `ipfs` table function which would check the required `(EXTERNAL, ipfs, READ)` permissions. (Note that the "ipfs" in the resource action is the name of the input source, not the table function.)

In another scenario, I might write a specialized table function to access an application's staging area. That staging area is on S3, but the user need not know that. (We might later change it to GCP.) In this case, the function would create a "virtual" input source, say, "abc-staging" and permissions would be granted on that name: `(EXTERNAL abc-staging, READ)`. The extension could use S3 internally, but that's an implementation detail.

## Code Revisions

Because we are adding a security change, we must ensure _all_ of Druid honors the new rules, not just MSQ. An outline of the changes:

* Modify each existing SQL table function to use `(EXTERNAL, <input source>, READ)` instead of `(EXTERNAL, EXTERNAL, READ)`. Note: this could be a breaking change for existing users.
* Extend the catalog code to add the input source security check in addition to the external table check, as noted above.
* Extend native batch ingest specs that use input sources to apply the new security checks
* Determine how to apply the rules to native batch ingest that use firehose factories
* Determine how to apply the rules to Hadoop ingest specs that use Hadoop FS paths
* Determine how to apply the rules to various flavors of realtime ingest

## Documentation and Release Notes

Documentation must explain the new permissions and security models. (Start with the text above.)

Release notes should announce when this model is available. The note must clearly state that if customers are granting permissions on the existing MSQ `(EXTERNAL, EXTERNAL, READ)` resource action, the rules should instead grant access on `(EXTERNAL, * READ)` to avoid ingestion failures.

### PR DESCRIPTION
**Pull request by zachjsh**
Partially Fixes #13837.

### Description

This change introduces the concept of input source type security model, proposed in #13837.. With this change, this feature is only available at the SQL layer, but we will expand to native layer in a follow up PR.

To enable this feature, the user must set the following property to true:

`druid.auth.enableInputSourceSecurity=true`

The default value for this property is false, which will continue the existing functionality of having the usage all external sources being authorized against the hardcoded resource action

`new ResourceAction(new Resource(ResourceType.EXTERNAL, ResourceType.EXTERNAL), Action.READ`

When this config is enabled, the users will be required to be authorized for the following resource action

`new ResourceAction(new Resource(ResourceType.EXTERNAL, {INPUT_SOURCE_TYPE}, Action.READ`

where `{INPUT_SOURCE_TYPE}` is the type of the input source being used;, http, inline, s3, etc..

Documentation has not been added for the feature as it is not complete at the moment, as we still need to enable this for the native layer in a follow up pr.

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: sql/src/main/java/org/apache/druid/sql/calcite/external/ExternalTableMacro.java
Comment (zachjsh) [inline sql/src/main/java/org/apache/druid/sql/calcite/external/ExternalTableMacro.java:1]: Can this class be removed? Doesn't seem to be used anywhere.

File: extensions-core/s3-extensions/src/test/java/org/apache/druid/catalog/model/table/S3InputSourceDefnTest.java [hunk 1]
```diff
@@ -345,6 +345,7 @@ public void testAdHocUri()
         CatalogUtils.stringListToUriList(uris),
         s3InputSource.getUris()
     );
+    assertEquals(Collections.singleton(S3StorageDruidModule.SCHEME), externSpec.inputSourceTypes);
```
Comment (paul-rogers) [inline extensions-core/s3-extensions/src/test/java/org/apache/druid/catalog/model/table/S3InputSourceDefnTest.java:348]: While this is OK, it muddies the semantics. There should be only one input source type, not a list. Yes, a list with one item is the same as a single value, but the list suggests that there could be more values, which there cannot be.

Second, the input source type should be the value returned from `typeValue()` that happens to be the same as the `SCHEME` constant, but that is an implementation detail.

Actually, the core problem here may be the lack of a `TYPE_KEY` constant in `S3InputSourceDefn`. Go ahead and add one, then use it here:

```java
public class S3InputSourceDefn extends FormattedInputSourceDefn
{
  public static final String TYPE_KEY = S3StorageDruidModule.SCHEME;
...

  @Override
  public String typeValue()
  {
    return TYPE_KEY;
  }
`

File: server/src/main/java/org/apache/druid/catalog/model/table/BaseInputSourceDefn.java [hunk 2]
```diff
@@ -151,7 +152,8 @@ protected ExternalTableSpec convertArgsToTable(
     return new ExternalTableSpec(
         convertArgsToSource(args, jsonMapper),
         convertArgsToFormat(args, columns, jsonMapper),
-        Columns.convertSignature(columns)
+        Columns.convertSignature(columns),
+        Collections.singleton(typeValue())
```
Comment (paul-rogers) [inline server/src/main/java/org/apache/druid/catalog/model/table/BaseInputSourceDefn.java:156]: Again, there can only ever be a single value.

File: server/src/main/java/org/apache/druid/catalog/model/table/ExternalTableSpec.java [hunk 3]
```diff
@@ -36,14 +37,17 @@
   public final InputSource inputSource;
   public final InputFormat inputFormat;
   @Nullable public final RowSignature signature;
+  public final Set<String> inputSourceTypes;
```
Comment (paul-rogers) [inline server/src/main/java/org/apache/druid/catalog/model/table/ExternalTableSpec.java:40]: Again, there can only ever be a single value.

File: server/src/main/java/org/apache/druid/server/security/AuthConfig.java [hunk 4]
```diff
@@ -97,6 +97,9 @@ public AuthConfig()
   @JsonProperty
   private final Set<String> securedContextKeys;
 
+  @JsonProperty
+  private final boolean inputSourceTypeSecurityEnabled;
```
Comment (paul-rogers) [inline server/src/main/java/org/apache/druid/server/security/AuthConfig.java:101]: Maybe `enableInputSourceSecurity` shorter. Starts with a verb.

File: sql/src/main/java/org/apache/druid/sql/calcite/external/DruidExternTableMacro.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+              .getOperandList()
+              .get(0))
+              .getValue())
+              .getValue();
+        }
+      }
+    }
+    // this shouldn't happen, as the sqlCall should have been validated by this point,
+    // and be guarenteed to have this parameter.
+    throw new RuntimeException("inputSource paramter not found in extern function");
```
Comment (paul-rogers) [inline sql/src/main/java/org/apache/druid/sql/calcite/external/DruidExternTableMacro.java:97]: paramter -> parameter

File: sql/src/main/java/org/apache/druid/sql/calcite/external/ExternalTableMacro.java
Comment (paul-rogers) [reply sql/src/main/java/org/apache/druid/sql/calcite/external/ExternalTableMacro.java:1]: Please remove it if it is not used.

File: sql/src/main/java/org/apache/druid/sql/calcite/external/Externals.java [hunk 6]
```diff
@@ -290,15 +294,19 @@ public static ExternalTable buildExternalTable(ExternalTableSpec spec, ObjectMap
...
   public static ResourceAction externalRead(String name)
   {
     return new ResourceAction(new Resource(name, ResourceType.EXTERNAL), Action.READ);
   }
 
-  public static ExternalTable toExternalTable(ExternalTableSpec spec, ObjectMapper jsonMapper)
+  public static ExternalTable toExternalTable(
+      ExternalTableSpec spec,
+      ObjectMapper jsonMapper,
+      Set<String> inputSourceTypes
```
Comment (paul-rogers) [inline sql/src/main/java/org/apache/druid/sql/calcite/external/Externals.java:308]: There will only ever be one input source type.

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteInsertDmlTest.java [hunk 7]
```diff
@@ -351,6 +373,94 @@ public void testInsertFromExternalWithSchema()
...
+        .expectResources(dataSourceWrite("dst"), externalRead("inline"))
+        .expectQuery(
+            newScanQueryBuilder()
+                .dataSource(externalDataSource)
+                .intervals(querySegmentSpec(Filtration.eternity()))
+                .columns("x", "y", "z")
+                .context(PARTITIONED_BY_ALL_TIME_QUERY_CONTEXT)
+                .build()
+        )
+        .expectLogicalPlanFrom("insertFromExternal")
```
Comment (paul-rogers) [inline sql/src/test/java/org/apache/druid/sql/calcite/CalciteInsertDmlTest.java:460]: Should we include a test in which authorization fails?

**Review by paul-rogers — CHANGES_REQUESTED** [review]
Thanks for tackling this task!

File: extensions-core/s3-extensions/src/test/java/org/apache/druid/catalog/model/table/S3InputSourceDefnTest.java [hunk 8]
```diff
@@ -695,14 +706,15 @@ public void testTableSpecAsConnection()
     Map<String, Object> args = new HashMap<>();
     args.put(S3InputSourceDefn.PATHS_PARAMETER, Collections.singletonList("bar/file.csv"));
     args.put(FormattedInputSourceDefn.FORMAT_PARAMETER, CsvFormatDefn.TYPE_KEY);
-    ExternalTableSpec externSpec = fn.apply("x", args, COLUMNS, mapper);
+    ExternalTableSpec externSpec = fn.apply("x", args, COLUMNS, mapper);;
```
Comment (paul-rogers) [inline extensions-core/s3-extensions/src/test/java/org/apache/druid/catalog/model/table/S3InputSourceDefnTest.java:709]: Nit: extra semicolon. Here and below. A static check probably complained.

**Review by paul-rogers — APPROVED** [review]
LGTM.

A couple of nits, and ensure there is a clean build, else +1.

**Review by abhishekrb19 — COMMENTED** [review]
LGTM. Just a few minor suggestions. Thanks!

**Comment by zachjsh** [comment]
thanks @abhishekrb19 ! I will take care of the minor comments in follow up pr.

**Pull request merged**